updates for ebpf / full host profiling (#29270)
DataDog/documentation: content/en/profiler/enabling/ebpf.md

---

> title: Enabling the eBPF Profiler

Author: yep - lets do it.

## Requirements

For a summary of the minimum and recommended runtime and tracer versions across all languages, read [Supported Language and Tracer Versions][7].

Supported operating systems
: Linux

Supported architecture
: `amd64` or `arm64` processors

Serverless
: `ebpf` is not supported on serverless platforms, such as AWS Lambda.

OS Settings
: TBD

Debugging information
: Symbols should be available locallly or can be  uploaded in CI via `datadog-ci` 

## Installation

The profiler can be used as a standalone executable.

### Running the Profiler
If the host is running workloads inside containers, it is recommended to run the profiler inside a container as well. A container image is [available][5].

If you're using Kubernetes, follow the instruction for [running in Kubernetes][8].

If you're directly using Docker, follow the instrcutins for [running in Docker][9].

If you're not using a container runtime, follow the instructions for running [directly on the host][10]

> ### Configuring the Profiler

Reviewer: ```suggestion
### Configuration
```

#### Local symbol upload (Experimental)
For compiled languages (C/C++/Rust/Go/...), the profiler can upload local symbols (when available) to Datadog for symbolication. Symbols need to be available locally (unstripped binaries).

To enable local symbol upload:

1. Set the `DD_HOST_PROFILING_EXPERIMENTAL_UPLOAD_SYMBOLS` environment variable to true.

> 2. Provide a Datadog API key through the `DD_API_KEY` environment variable.

Reviewer: ```suggestion
2. Provide a [Datadog API key][11] through the `DD_API_KEY` environment variable.
```

> 3. Provide a Datadog APP key through the `DD_APP_KEY` environment variable.

Reviewer: ```suggestion
3. Provide a [Datadog application key][13] through the `DD_APP_KEY` environment variable.
```

> 4. Set the `DD_SITE environment` variable to your Datadog site (e.g. datadoghq.com, datadoghq.eu, us5.datadoghq.com, ...).

Reviewer: ```suggestion
4. Set the `DD_SITE environment` variable to your [Datadog site][12]. Your site is: {{< region-param key="dd_site" code="true" >}}
```
Note this will auto-populate based on the site they have selected.

### Build

> You must first ensure you have the correct version of go installed. In order to build the profiler directly on your machine, you can simply run:
> 
>    ```shell
>    make
>    ```

Reviewer: `````suggestion
To build the Full-Host Profiler directly on your machine:

1. Ensure you have the required version of Go installed:
   ```shell
   go version
   ```
2. Run:
   ```shell
   make
   ```
`````

We should note what the required version of Go is in the Requirements section. Do we know that yet?

### Library
TBD

### Deploying the shared library

TBD

#### Installing the library

TBD
#### Appending a search directory

TBD

## Configuration

TBD

> ## Not sure what to do next?

Reviewer: ```suggestion
## What's next?
```

> The [Getting Started with Profiler][6] guide takes a sample service with a performance problem and shows you how to use Continuous Profiler to understand and fix the problem.

Reviewer: ```suggestion
After installing the Full-Host Profiler, see the [Getting Started with Profiler][6] to learn how to use Continuous Profiler to identify and fix performance problems.
```

> ## Further Reading

Reviewer: ```suggestion
## Further reading
```

{{< partial name="whats-next/whats-next.html" >}}

> [2]: shttps://github.com/DataDog/dd-otel-host-profiler/releases/

Reviewer: ```suggestion
[2]: https://github.com/DataDog/dd-otel-host-profiler/releases/
```

[3]: https://app.datadoghq.com/profiling
[4]: /getting_started/tagging/unified_service_tagging
[5]: https://github.com/DataDog/dd-otel-host-profiler/pkgs/container/dd-otel-host-profiler/.
[6]: /getting_started/profiler/
[7]: /profiler/enabling/supported_versions/
[8]: https://github.com/DataDog/dd-otel-host-profiler/blob/main/doc/running-in-kubernetes.md
[9]: https://github.com/DataDog/dd-otel-host-profiler/blob/main/doc/running-in-docker.md

> [10]: https://github.com/DataDog/dd-otel-host-profiler/blob/main/doc/running-on-host.md

Reviewer: ```suggestion
[10]: https://github.com/DataDog/dd-otel-host-profiler/blob/main/doc/running-on-host.md
[11]: https://app.datadoghq.com/organization-settings/api-keys
[12]: /getting_started/site/
[13]: https://app.datadoghq.com/access/application-keys
```
Adding links for suggested changes in Configuration section.

---
title: Enabling the eBPF Profiler
code_lang: ebpf
type: multi-code-lang
code_lang_weight: 90
aliases:
  - /tracing/profiler/enabling/ebpf/
further_reading:
    - link: 'getting_started/profiler'
      tag: 'Documentation'
      text: 'Getting Started with Profiler'
    - link: 'profiler/profiler_troubleshooting/ebpf'
      tag: 'Documentation'
      text: 'Fix problems you encounter while using the profiler'
---

<div class="alert alert-warning">
<code>ebpf</code> is in beta. Datadog recommends evaluating the profiler in a non-sensitive environment before deploying in production.
</div>

The eBPF profiler is our eBPF-based host profiler built on OpenTelemetry and designed to send profiling data to the Datadog backend via the Datadog Agent. It supports symbolication for compiled languages and is optimized for containerized environments like Docker and Kubernetes.

## Requirements

For a summary of the minimum and recommended runtime and tracer versions across all languages, read [Supported Language and Tracer Versions][7].

Supported operating systems

> : Linux (glibc or musl)

Reviewer 2: ```suggestion
: Linux
```

Supported architecture
: `amd64` or `arm64` processors

Serverless
: `ebpf` is not supported on serverless platforms, such as AWS Lambda.

OS Settings

> : `perf_event_paranoid` kernel setting is 2 or less (see [Troubleshooting][1])

Reviewer 2: This is not relevant for the eBPF profiler, we do need though some OS permissions. Let's flag this as TBD for now?

Author: can do

Debugging information

> : Symbols should be available. The profiler cannot provide human-readable function names if the symbol table is stripped.

Reviewer 2: This is not the case since we support remote symbolication. Symbols should either:
* Be available locally (recommended)
* Be uploaded in CI via (datadog-ci) (we likely want a dedicated documentation page for this?)

## Installation

> The profiler can be used either as a standalone executable or as a library. Skip to [library installation instructions](#library) if you want to use it as a library.

Reviewer 2: ```suggestion
The profiler can be used as a standalone executable.
```

If you're directly using Docker, follow the instrcutins for [running in Docker][9].

If you're not using a container runtime, follow the instructions for running [directly on the host][10]

### Configuring the Profiler
#### Local symbol upload (Experimental)
For compiled languages (C/C++/Rust/Go/...), the profiler can upload local symbols (when available) to Datadog for symbolication. Symbols need to be available locally (unstripped binaries).

To enable local symbol upload:

1. Set the `DD_HOST_PROFILING_EXPERIMENTAL_UPLOAD_SYMBOLS` environment variable to true.
2. Provide a Datadog API key through the `DD_API_KEY` environment variable.
3. Provide a Datadog APP key through the `DD_APP_KEY` environment variable.
4. Set the `DD_SITE environment` variable to your Datadog site (e.g. datadoghq.com, datadoghq.eu, us5.datadoghq.com, ...).

### Build
You must first ensure you have the correct version of go installed. In order to build the profiler directly on your machine, you can simply run:

   ```shell
   make
   ```

### Library
TBD

### Deploying the shared library

TBD

#### Installing the library

TBD
#### Appending a search directory

> TBD

Reviewer 2: Let's remove all this since we won't support this method.

## Configuration

> The `environment`, `service`, and `service_version` settings are recommended, as they are used by the Profiling UI.

Reviewer 2: We might want to put this as TBD as well, until we've figured out the right configuration bits (e.g. with the current planned work, `service` is likely not relevant)

See the [full list of parameters][5] or use the command line.

```bash
ebpf --help
```

### Logging

You can configure logging to one of several endpoints:
- `stdout` prints the logs to standard output stream (the default).
- `stderr` prints the logs to the standard error stream.
- `syslog` publishes the logs to syslog, attempting to adhere to the specification in RFC 3164.
- `disable` disables the logs entirely.

> - Any other value is treated as a file path, with a leading `/` designating an absolute path.

Reviewer 2: For full host we only support two verbose logging flags, one for the profiler, and one for eBPF https://github.com/DataDog/dd-otel-host-profiler/blob/5fcfce395db9222b9702fb9cb4aeb92239e45e04/cli_flags.go#L267-L281 I'm not sure we want to document this though.

Author: I'll remove it

### Global

If you want to instrument all running process, you can try out the `--global` option.
Global mode is intended for debug purposes. This requires elevated permissions. Depending on your setup, this can mean running as root, granting `CAP_PERFMON`, `CAP_SYSADMIN`, or setting `perf_event_paranoid` to `-1`.

```bash
./ebpf --environment staging --global --service_version full-host-profile
```

> For most configurations, this consists of all processes visible within the profiler's PID namespace.

Reviewer 2: I believe this section is mostly relevant to ddprof, and not the eBPF profiler.
